src/renderers/shared/fiber/ReactFiberScheduler.js

```diff
@@ -492,63 +511,113 @@ module.exports = function<T, P, I, TI, C>(config : HostConfig<T, P, I, TI, C>) {
   }
 
   function workLoop(priorityLevel) {
-    // If there are errors, use a forked version of performUnitOfWork
-    // TODO: Need to make commitAllWork throw on error before doing this
-    if (true || (capturedErrors && capturedErrors.size)) {
+    outer: while (true) {
+      if (!nextUnitOfWork) {
+        nextUnitOfWork = findNextUnitOfWork();
+      }
+
+      // If there are errors, use a forked version of performUnitOfWork
+      // TODO: Need to make commitAllWork throw on error before doing this
+      if (capturedErrors && capturedErrors.size) {
+        while (true) {
+          if (!nextUnitOfWork && !pendingCommit) {
+            nextUnitOfWork = findNextUnitOfWork();
+          }
+          // If we're currently flushing Task work, but the pending commit does
+          // not have Task priority, don't commit it — we're at the end of a
+          // deferred work batch and the commit should happen in the next batch.
+          if (
+            pendingCommit &&
+            !(priorityLevel === TaskPriority && pendingCommit.pendingWorkPriority !== TaskPriority)
+          ) {
+            commitAllWork(pendingCommit);
+          } else if (nextUnitOfWork && nextPriorityLevel !== NoWork && nextPriorityLevel <= priorityLevel) {
+            if (capturedErrors && capturedErrors.has(nextUnitOfWork)) {
+              nextUnitOfWork = performFailedUnitOfWork(nextUnitOfWork);
+            } else {
+              nextUnitOfWork = performUnitOfWork(nextUnitOfWork);
+            }
+          } else {
+            return;
+          }
+        }
+      }
+
+      // Otherwise, handle the normal, faster case where there are no errors:
       while (true) {
         if (!nextUnitOfWork && !pendingCommit) {
           nextUnitOfWork = findNextUnitOfWork();
         }
-        if (pendingCommit && !(priorityLevel === TaskPriority && pendingCommit.pendingWorkPriority !== TaskPriority)) {
-          nextUnitOfWork = commitAllWork(pendingCommit);
-        } else if (nextUnitOfWork && nextPriorityLevel !== NoWork && nextPriorityLevel <= priorityLevel) {
-          if (capturedErrors && capturedErrors.has(nextUnitOfWork)) {
-            nextUnitOfWork = performFailedUnitOfWork(nextUnitOfWork);
-          } else {
-            nextUnitOfWork = performUnitOfWork(nextUnitOfWork);
+        // If we're currently flushing Task work, but the pending commit does
+        // not have Task priority, don't commit it — we're at the end of a
+        // deferred work batch and the commit should happen in the next batch.
+        if (
+          pendingCommit &&
+          !(priorityLevel === TaskPriority && pendingCommit.pendingWorkPriority !== TaskPriority)
```

Comment:
Can you extract this pattern to a boolean variable please? Double negation is hard to understand.